// @gabriele-salvo/socket-chat-angular/esm2022/lib/services/chat.service.mjs
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { io } from 'socket.io-client';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class ChatService {
    constructor(http) {
        this.http = http;
        this.baseUrl = 'https://0e9b-93-35-217-254.ngrok-free.app';
        const room = localStorage.getItem('chatRoom');
        this.socket = io('https://0e9b-93-35-217-254.ngrok-free.app');
    }
    getGlobalMessages() {
        return this.http.get(`${this.baseUrl}/global-messages`);
    }
    sendMessage(message) {
        this.socket.emit('sendMessage', message);
    }
    getMessages() {
        return new Observable((observer) => {
            this.socket.on('message', (data) => {
                observer.next(data);
            });
        });
    }
    joinRoom(room, username) {
        this.socket.emit('joinRoom', { room, username: username || 'default' });
    }
    leaveRoom(room, username) {
        this.socket.emit('leaveRoom', { room, username: username || 'default' });
    }
    sendMessageToRoom(room, message, username) {
        this.socket.emit('sendMessage', { room, message, username });
    }
    generateRoomCode() {
        return Math.random().toString(36).substring(2, 9);
    }
    getActiveRooms() {
        return this.http.get('http://localhost:3000/active-rooms');
    }
    getActiveRoomsUpdates() {
        return new Observable((observer) => {
            this.socket.on('activeRooms', (rooms) => {
                observer.next(rooms);
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChatService, deps: [{ token: i1.HttpClient }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChatService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChatService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.HttpClient }]; } });
