// @gabriele-salvo/socket-chat-angular/esm2022/lib/chat-lib.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { ChatComponent } from './chat/chat.component';
import { ChatRoomComponent } from './chat-room/chat-room.component';
import { ChatService } from './services/chat.service';
import { WrapperComponent } from './wrapper/wrapper.component';
import * as i0 from "@angular/core";
export class ChatLibModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChatLibModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: ChatLibModule, declarations: [ChatComponent, ChatRoomComponent, WrapperComponent], imports: [CommonModule, FormsModule, HttpClientModule], exports: [ChatComponent, ChatRoomComponent, WrapperComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChatLibModule, providers: [ChatService], imports: [CommonModule, FormsModule, HttpClientModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChatLibModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ChatComponent, ChatRoomComponent, WrapperComponent],
                    imports: [CommonModule, FormsModule, HttpClientModule],
                    exports: [ChatComponent, ChatRoomComponent, WrapperComponent],
                    providers: [ChatService],
                }]
        }] });
