// @gabriele-salvo/socket-chat-angular/esm2022/lib/chat-room/chat-room.component.mjs
import { Component, Input, Output, EventEmitter } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../services/chat.service";
import * as i2 from "@angular/common";
import * as i3 from "@angular/forms";
export class ChatRoomComponent {
    constructor(chatService) {
        this.chatService = chatService;
        this.room = '';
        this.username = 'default';
        this.message = '';
        this.messages = [];
        this.inRoom = false;
        this.leave = new EventEmitter();
    }
    ngOnInit() {
        if (this.room === 'Global Chat') {
            this.fetchGlobalMessages();
        }
        this.chatService.getMessages().subscribe((data) => {
            if (data.room === this.room) {
                this.messages.push(data);
            }
        });
    }
    fetchGlobalMessages() {
        this.chatService.getGlobalMessages().subscribe((messages) => {
            this.messages = messages.filter((msg) => msg.room === 'Global Chat');
        });
    }
    joinRoom() {
        this.chatService.joinRoom(this.room, this.username || 'default');
        this.inRoom = true;
    }
    leaveRoom() {
        console.log(`Leaving room: ${this.room}`);
        this.chatService.leaveRoom(this.room, this.username || 'default');
        this.inRoom = false;
        this.leave.emit();
    }
    sendMessage() {
        if (this.username != null) {
            this.chatService.sendMessageToRoom(this.room, this.message, this.username);
        }
        this.message = '';
    }
    generateRoomCode() {
        this.room = Math.random().toString(36).substring(2, 9);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChatRoomComponent, deps: [{ token: i1.ChatService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: ChatRoomComponent, selector: "app-chat-room", inputs: { room: "room", username: "username" }, outputs: { leave: "leave" }, ngImport: i0, template: "<div class=\"chat-room\">\n  <h3 class=\"title is-5 has-background-primary has-text-white p-3\">\n    {{ room }}\n  </h3>\n  <div class=\"messages\">\n    <div class=\"message\" *ngFor=\"let msg of messages\">\n      <strong>{{ msg.username }}:</strong> {{ msg.message }}\n    </div>\n  </div>\n\n  <div class=\"field has-addons\">\n    <div class=\"control is-expanded\">\n      <input class=\"input\" [(ngModel)]=\"message\" placeholder=\"Type a message\" />\n    </div>\n    <div class=\"control\">\n      <button class=\"button is-info\" (click)=\"sendMessage()\">Send</button>\n    </div>\n  </div>\n\n  <button class=\"button is-danger mr-5\" (click)=\"leaveRoom()\">\n    Leave Room\n  </button>\n</div>\n", styles: [".chat-room{display:flex;flex-direction:column;justify-content:space-between;height:80vh;width:30vw;background-color:#fff;border:1px solid #ddd;padding:.75rem;overflow:hidden;position:fixed;bottom:20px;right:20px;box-shadow:0 2px 10px #0000001a;z-index:1000}.messages{overflow-y:auto;flex-grow:1}.message{margin-bottom:.5rem;padding:.5rem;background-color:#f5f5f5;border-radius:4px}\n"], dependencies: [{ kind: "directive", type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i3.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { kind: "directive", type: i3.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i3.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChatRoomComponent, decorators: [{
            type: Component,
            args: [{ selector: 'app-chat-room', template: "<div class=\"chat-room\">\n  <h3 class=\"title is-5 has-background-primary has-text-white p-3\">\n    {{ room }}\n  </h3>\n  <div class=\"messages\">\n    <div class=\"message\" *ngFor=\"let msg of messages\">\n      <strong>{{ msg.username }}:</strong> {{ msg.message }}\n    </div>\n  </div>\n\n  <div class=\"field has-addons\">\n    <div class=\"control is-expanded\">\n      <input class=\"input\" [(ngModel)]=\"message\" placeholder=\"Type a message\" />\n    </div>\n    <div class=\"control\">\n      <button class=\"button is-info\" (click)=\"sendMessage()\">Send</button>\n    </div>\n  </div>\n\n  <button class=\"button is-danger mr-5\" (click)=\"leaveRoom()\">\n    Leave Room\n  </button>\n</div>\n", styles: [".chat-room{display:flex;flex-direction:column;justify-content:space-between;height:80vh;width:30vw;background-color:#fff;border:1px solid #ddd;padding:.75rem;overflow:hidden;position:fixed;bottom:20px;right:20px;box-shadow:0 2px 10px #0000001a;z-index:1000}.messages{overflow-y:auto;flex-grow:1}.message{margin-bottom:.5rem;padding:.5rem;background-color:#f5f5f5;border-radius:4px}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.ChatService }]; }, propDecorators: { room: [{
                type: Input
            }], username: [{
                type: Input
            }], leave: [{
                type: Output
            }] } });
